### PR TITLE
add metrics for jobs waiting when OOM

### PR DESCRIPTION
**Pull request by chengren311**
monitor the number of waiting jobs when OOM

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 1]
```diff
@@ -71,4 +73,12 @@ public void markFlowFail() {
   public void setDBConnectionTime(long milliseconds) {
     dbConnectionTime.set(milliseconds);
   }
+
+  /**
+   * Mark the occurrence of an job waiting event due to OOM
+   */
+  public void markJobWait() {
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: This name should be specific for OOM, right?

Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: updated

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java [hunk 2]
```diff
@@ -87,10 +88,13 @@ public void run() throws Exception {
           break;
         }
         if (attempt < ServerInternals.MEMORY_CHECK_RETRY_LIMIT) {
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: One thing still confusing me is that There looks duplicate checking logics in Line90 and line 84.
`attempt <= ServerInternals.MEMORY_CHECK_RETRY_LIMIT`

Are they really necessary?

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java [hunk 3]
```diff
@@ -87,10 +88,13 @@ public void run() throws Exception {
...
-          info(String.format(oomMsg + ", sleep for %s secs and retry, attempt %s of %s", TimeUnit.MILLISECONDS.toSeconds(ServerInternals.MEMORY_CHECK_INTERVAL), attempt, ServerInternals.MEMORY_CHECK_RETRY_LIMIT));
+          info(String.format(oomMsg + ", sleep for %s secs and retry, attempt %s of %s", TimeUnit.MILLISECONDS.toSeconds(ServerInternals.MEMORY_CHECK_INTERVAL_MS), attempt, ServerInternals.MEMORY_CHECK_RETRY_LIMIT));
+          if (attempt == 1) {
+            CommonMetrics.INSTANCE.markOOMJobWait();
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: Is it better to track the current number of jobs blocked waiting for memory? This will also give us the information about how long the jobs wait.

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java (hunk 2 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: we just want to skip waiting for last attempt

File: azkaban-common/src/main/java/azkaban/constants/ServerInternals.java [hunk 4]
```diff
@@ -29,7 +29,7 @@
 
 
   // Memory check retry interval when OOM in ms
-  public static final long MEMORY_CHECK_INTERVAL = 1000*60*1;
+  public static final long MEMORY_CHECK_INTERVAL_MS = 1000 * 60 * 1;
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/constants/ServerInternals.java]: will update the description

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 5]
```diff
@@ -31,6 +31,7 @@
 
   private Meter dbConnectionMeter;
   private Meter flowFailMeter;
+  private Meter OOMwaitingJobMeter;
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: This requires moving ProcessJob and all its dependent class to executor package. I think it won't be that complicated since we can just move the files but keep the package name unchanged. But it's definitely a big change which should be in another commit.

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java [hunk 6]
```diff
@@ -100,13 +103,15 @@ public void run() throws Exception {
...
           if(killed) {
+            CommonMetrics.INSTANCE.unmarkOOMJobWait();
             info(String.format("Job %s was killed while waiting for memory check retry", getId()));
             return;
           }
         }
       }
 
       if (!isMemGranted) {
+        CommonMetrics.INSTANCE.unmarkOOMJobWait();
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: Should we also have a logger.error message here besides an Exception.

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java (hunk 3 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: updated

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java (hunk 6 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: good idea, updated

File: azkaban-common/src/main/java/azkaban/constants/ServerInternals.java (hunk 4 above)
Comment (shuzhang1989) [inline azkaban-common/src/main/java/azkaban/constants/ServerInternals.java]: +1 for the good practice that one diff only does one thing

Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/constants/ServerInternals.java]: Consider using Duration class instead?

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 7]
```diff
@@ -43,6 +43,7 @@
   private void setupAllMetrics() {
     dbConnectionMeter = MetricsUtility.addMeter("DB-Connection-meter", registry);
     flowFailMeter = MetricsUtility.addMeter("flow-fail-meter", registry);
+    MetricsUtility.addGauge("OOM-waiting-job-meter", registry, OOMWaitingJobCount::get);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: This is a gauge not a meter. Will OOM-waiting-job-count be a better name?

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 8]
```diff
@@ -71,4 +72,20 @@ public void markFlowFail() {
   public void setDBConnectionTime(long milliseconds) {
     dbConnectionTime.set(milliseconds);
   }
+
+  /**
+   * Mark the occurrence of an job waiting event due to OOM
+   */
+  public void markOOMJobWait() {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: Similarly, would it be better to rename it " incrementOOMJobWaitCount"?

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 9]
```diff
@@ -71,4 +72,20 @@ public void markFlowFail() {
...
+    OOMWaitingJobCount.incrementAndGet();
+  }
+
+  /**
+   * Unmark the occurrence of an job waiting event due to OOM
+   */
+  public void unmarkOOMJobWait() {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: ditto

File: azkaban-common/src/main/java/azkaban/constants/ServerInternals.java (hunk 4 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/constants/ServerInternals.java]: yup, but will create another PR refactoring all of them.

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java (hunk 7 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: update

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java (hunk 8 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: updated

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java (hunk 9 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: updated

**Comment by HappyRay** [comment]
LGTM. I hope you can move this logic to executor as discussed earlier.
